tests/test_request_handling.py

```diff
@@ -0,0 +1,554 @@
+            "Invalid value for 'function_call': 'function_call' is only allowed when 'functions' are specified.",
+            "Edge case handling failed: function_call provided without providing functions",
+        )
+
+        # When tool_choice is provided and tools is not provided
+        request = ChatCompletionRequest(
+            model=self.served_model,
+            messages=self.default_messages,
+            functions=self.default_functions,
+            tool_choice="none",
+        )
+        response = convert_jsonresponse_to_json(
+            await check_all_errors(request=request, served_model=self.served_model)
+        )
+        self.assertEqual(
+            response["message"],
+            "Invalid value for 'tool_choice': 'tool_choice' is only allowed when 'tools' are specified.",
+            "Edge case handling failed: tool_choice provided without providing tools",
+        )
+
+    async def test_request_handling(self):
+        def generate_raw_response(
+            gen_text: bool,
+            num_tool_calls: Optional[int],
+            tool_func_choice: Union[str, Tool, Function],
+        ):
+            messages = []
+            if gen_text:
+                messages.append({"role": "assistant", "content": self.default_text_str})
+            if num_tool_calls is not None:
+                tool_calls = []
+                for tool_call_args in self.default_tool_call_args[:num_tool_calls]:
+                    tool_calls.append(
+                        {
+                            "type": "function",
+                            "function": {
+                                "name": self.default_tool_call_name,
+                                "arguments": tool_call_args,
+                            },
+                        }
+                    )
+                messages.append({"role": "assistant", "tool_calls": tool_calls})
+
+            raw_response = "".join(
+                [
+                    prompt_template.convert_message_to_prompt(message)
+                    for message in messages
+                ]
+            ).rstrip()
+            null_response = prompt_template.convert_message_to_prompt(
+                {"role": "assistant"}
+            )
+            raw_response = raw_response[len(null_response) :]
+            for stop_token in prompt_template.get_stop_tokens_for_generation():
+                raw_response = raw_response.replace(stop_token, "")
+
+            if prompt_template.version == "v2":
+                if tool_func_choice == "none":
+                    raw_response = raw_response[len("all\n<|content|>") :]
+                elif isinstance(tool_func_choice, Tool) or isinstance(
+                    tool_func_choice, Function
+                ):
+                    raw_response = raw_response[
+                        len(f"{self.default_tool_call_name}\n<|content|>") :
+                    ]
+            elif prompt_template.version == "v2.llama3":
```

Comment:
I mean we don't need to check if prompt_template_version == a specific_version. If not when we add a new template (for example template of Phi3), we have to write if-else here. I think we can implement this without if-else for prompt_template.version using abstract methods:
+ convert_message_to_prompt
+ get_force_text_generation_prefix
+ get_force_function_call_prefix